### PR TITLE
adding an event cache to handle burst of requests generating multiple…

### PR DESCRIPTION
**Pull request by aadc-dev**
adding an event cache to handle burst of requests generating multiple access requests. Via a first requests "wins" - subsequent requests within a provided TTL will be rejected and debounced. EventCache implements "clean up on add" semantics, removing expired entries on TryAdd calls.

To fix: https://github.com/gravitational/shared-workflows/issues/413
Cloud-terraform PR adding `event_cache_ttl ` config entry: https://github.com/gravitational/cloud-terraform/pull/7077

### CODE REVIEW

**Review by fheinecke — APPROVED** [review]
Looks reasonable to me provided that this has been tested end to end

File: tools/approval-service/internal/service/ttleventcache.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+		if ttl > 4*time.Second {
+			return ttl / 4
+		}
+		return 1 * time.Second
+	}
+
+	c := &TTLEventCache{
+		items:           make(map[string]time.Time),
+		ttl:             ttl,
+		cleanupInterval: defaultCleanupInterval(ttl),
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache.go]: this can be simplified to `cleanupInterval: max(ttl / 4, 1*time.Second)`, removing the inline function above.

File: tools/approval-service/internal/service/ttleventcache.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+// evictor runs in a background goroutine and periodically removes expired entries from the cache.
+func (c *TTLEventCache) evictor() {
+	ticker := time.NewTicker(c.cleanupInterval)
+	defer ticker.Stop()
+	defer close(c.done)
+
+	for {
+		select {
+		case <-ticker.C:
+			now := time.Now()
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache.go]: The ticker sends the current time each time it fires, so you can do `case now := <-ticker.C` and remove this line

File: tools/approval-service/internal/service/ttleventcache.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	}
+
+	if expiry, ok := c.items[eventID]; ok {
+		// unexpired -> reject
+		if now.Before(expiry) {
+			return false
+		}
+
+		// expired -> delete and allow insertion
+		delete(c.items, eventID)
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache.go]: no need to delete here. insertion will just overwrite

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	const goroutines = 500
+	var wg sync.WaitGroup
+
+	var wins int32
+	var mu sync.Mutex
+
+	// Act
+	for i := 0; i < goroutines; i++ {
+		wg.Go(func() {
+			if ec.TryAdd(eventIDForTest) {
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: There is no guarantee that all of the calls to `TryAdd()` will happen within 200ms (the ttl). On a slow CI runner with lots of contention, this test could be flaky.

Could you instead use the synctest package to test this behaviour more deterministically?

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	go func() {
+		done <- ec.TryAdd(eventIDInProgressForTest)
+	}()
+	addResult := <-done
+	if addResult {
+		t.Fatalf("expected concurrent TryAdd to be rejected while cooldown is active")
+	}
+
+	// after TTL elapses, TryAdd should succeed again
+	time.Sleep(ttl + 20*time.Millisecond)
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: Don't use `time.Sleep` in tests if you are not using [`synctest`](https://pkg.go.dev/testing/synctest)

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	if !ec.TryAdd(eventIDInProgressForTest) {
+		t.Fatalf("expected first TryAdd to succeed")
+	}
+
+	done := make(chan bool, 1)
+	go func() {
+		done <- ec.TryAdd(eventIDInProgressForTest)
+	}()
+	addResult := <-done
+	if addResult {
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: On a slow test runner, there's no guarantee that 100ms has not passed in which case a true result here would be valid. This makes the test flaky. Use `synctest` instead to have control over time in tests.

File: tools/approval-service/internal/service/ttleventcache.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		}
+	}
+
+	cleanup := func() error {
+		_ = c.Stop(context.Background())
+		return nil
+	}
+
+	// start evictor
+	go c.evictor()
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache.go]: There does not really seem to be a need for the evictor to run concurrently. It adds a lot of complexity that is hard to read/comprehend/verify and difficult to test. Instead you can just expire entries as the cache is looked up in `TryAdd()` and do it synchronously. No locking needed, no concurrency, no stopping the cache for cleanup.

It can be made reasonably efficient by keeping an ordered list of expiries as well as the map. On each TryAdd, iterate the expiry slice up to `now`, expiring all entries iterated. Remove them all from the slice. Then just do the map lookup and add an entry if necessary. No concurrency needed, no channels, no mutexes.

Comment (camscale) [reply tools/approval-service/internal/service/ttleventcache.go]: > No locking needed,

Although I realise now the cache may be called concurrently, so a lock for `TryAdd()` will probably be needed.

File: tools/approval-service/internal/approvalservice/approvalservice.go [hunk 8]
```diff
@@ -109,6 +109,11 @@ func NewFromConfig(ctx context.Context, cfg config.Root, opts ...Opt) (*Service,
 		return nil, fmt.Errorf("creating event processor: %w", err)
 	}
 	a.processor = processor
+	defer func() {
+		if err := processor.Close(); err != nil {
+			a.log.Warn("Failed to close event processor")
```
[the remark this answers is not in the document]
Comment (aadc-dev) [reply tools/approval-service/internal/approvalservice/approvalservice.go]: Right at this point, not much, the Close function for the approval service is meant to do service level cleanup in one location. As of this PR, it calls the Stop function on the cache to stop allowing more entries from being processed, and stops the evictor go routine.

File: tools/approval-service/internal/service/ttleventcache.go (hunk 7 above)
Comment (aadc-dev) [reply tools/approval-service/internal/service/ttleventcache.go]: Yup a lock on TryAdd would still be needed, and each call to TryAdd will potentially take longer, having do cleanup and keeping track of two data structures to perform deletion.

I am still loving delegating the cleanup to the evictor, to keep the TryAdd fast and simple. I guess we would be moving some complexity from one place to another.
Please let me know what you think and if we feel strongly about doing cleanup on TryCatch.

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+// exactly one caller wins (returns true) and others are rejected within the TTL
+func TestTryAddConcurrency(t *testing.T) {
+	// Arrange
+	ttl := 200 * time.Millisecond
+	ec, cleanupFunc, err := service.MakeTTLEventCache(ttl)
+	if err != nil {
+		t.Fatalf("MakeTTLCache error: %v", err)
+	}
+	defer func() {
+		_ = cleanupFunc
```
[the remark this answers is not in the document]
Comment (aadc-dev) [reply tools/approval-service/internal/service/ttleventcache_test.go]: yeah good catch.
I will also be removing the cleanup func, will be calling ttl.Stop from the service level Stop function

File: tools/approval-service/internal/service/ttleventcache.go (hunk 7 above)
Comment (camscale) [reply tools/approval-service/internal/service/ttleventcache.go]: I'm not particularly concerned with the efficiency of this - it is a service that is called so infrequently in the scheme of things that I don't think it matters, and the difference would be negligible anyway. I actually think the concurrent evictor is less efficient overall as it will be a ticker firing every 4 seconds largely doing no work as our release rate of approximately 2 releases per week on average means this service is mostly idle.

I am more concerned with simplicity and maintainability particularly as this service is critical to us getting releases out. And the complexity of expiring entries in-line during a `TryAdd()` seems to be to be a lot less than managing the concurrency of the evictor ensuring that the locking is always correct and the channel shutdown logic is correct. Not just now in this PR, but ongoing as the code evolves.

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+		cleanupTicker := 2 * time.Second
+
+		ec, err := service.NewTTLEventCache(ttl, service.WithCleanupInterval(cleanupTicker))
+		if err != nil {
+			t.Fatalf("MakeTTLCache error: %v", err)
+		}
+
+		// ensure evictor is stopped
+		defer func() {
+			ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: Consider using `t.Context()` instead of `context.Background()` in tests.

File: tools/approval-service/internal/service/ttleventcache.go [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+//
+// Returns true if the ID was added, or false if the ID is already in the cache and has not expired.
+//
+// Calls to TryAdd will prune any expired entries from the cache before performing the check.
+func (c *TTLEventCache) TryAdd(id string) bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := time.Now()
+	c.evictExpired(now)
```
Comment (fheinecke) [inline tools/approval-service/internal/service/ttleventcache.go]: This could potentially cause memory usage to balloon down the road under specific conditions if usage increased or changed. Can you add a comment here flagging this so that if we get hit by this in say three years, it's easier to identify what's going on?

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+		keyABCForTest := "ABC"
+		key123ForTest := "123"
+
+		isSuccessfulKeyOneFirstTry := ec.TryAdd(keyABCForTest)
+		if !isSuccessfulKeyOneFirstTry {
+			t.Errorf("TryAdd(keyOne) returned false on first add, expected true")
+		}
+
+		// wait for TTL to pass (virtual sleep)
+		time.Sleep(ttl + 10*time.Millisecond)
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: You do not need the +10ms here. You should be able to sleep the exact TTL due to being in a synctest bubble. This makes the testing more precise and you are able to test boundary conditions should they matter.

Same in the other tests.

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+	"sync/atomic"
+	"testing"
+	"testing/synctest"
+	"time"
+)
+
+func (c *TTLEventCache) checkInternalState(t *testing.T, expectedMapLen, expectedListLen int) {
+	t.Helper()
+	c.mu.Lock()
+	defer c.mu.Unlock()
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: This locking should not be needed. If it is, then the test is non-deterministic as it is obviously changing underneath the test check, meaning it is a flaky test. Instead if you do not take the lock the race detector should identify this as a problem if there is any concurrency of note that affects the tests.

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+
+		var successfulAdds int32
+
+		// Schedule one small step per logical request. The synctest harness runs these
+		// steps under a deterministic scheduler so inter-leavings are reproducible.
+		for i := 0; i < distinctKeys; i++ {
+			key := fmt.Sprintf("key-%d", i)
+			for j := 0; j < requestsPerKey; j++ {
+				if ec.TryAdd(key) {
+					atomic.AddInt32(&successfulAdds, 1)
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: Is this atomic needed? There doesn't seem to be any concurrency in the test or the unit under test. I'd also rename the "Concurrency" in the test to "Multiple" perhaps since there is no concurrency here.

File: tools/approval-service/internal/service/ttleventcache_test.go [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+		if !ec.TryAdd("B") {
+			t.Fatal("TryAdd(B) failed")
+		} // "B" expires at T=150ms
+
+		ec.checkInternalState(t, 2, 2)
+
+		// 3. Wait 60ms, time is now T=110ms
+		// "A" (T=100ms) is expired.
+		// "B" (T=150ms) is NOT expired.
+		time.Sleep(60 * time.Millisecond)
```
Comment (camscale) [inline tools/approval-service/internal/service/ttleventcache_test.go]: Consider sleeping 50ms-1 and checking that both are in the cache, and then sleeping 1 and checking the first expired. This tests the actual boundary and is doable with synctest.

**Comment by aadc-dev** [comment]
Tested this on Staging.
Thanks @doggydogworld

**Review by r0mant — APPROVED** [review]
